nexus/src/app/background/tasks/tuf_artifact_replication.rs

```diff
@@ -0,0 +1,941 @@
+                request_debug_ringbuf: self.request_debug_ringbuf.clone(),
+                local_repos: self.local.len(),
+            })
+            .unwrap()
+        }
+        .boxed()
+    }
+}
+
+impl ArtifactReplication {
+    pub fn new(
+        datastore: Arc<DataStore>,
+        local_rx: mpsc::Receiver<ArtifactsWithPlan>,
+    ) -> ArtifactReplication {
+        ArtifactReplication {
+            datastore,
+            local: Vec::new(),
+            local_rx,
+            request_debug_ringbuf: Arc::new(VecDeque::new()),
+            lifetime_counters: TufArtifactReplicationCounters::default(),
+        }
+    }
+
+    fn insert_debug_requests(
+        &mut self,
+        mut requests: Vec<TufArtifactReplicationRequest>,
+        counters: &mut TufArtifactReplicationCounters,
+    ) {
+        for request in &requests {
+            counters.inc(request);
+        }
+
+        // `Arc::make_mut` will either directly provide a mutable reference
+        // if there are no other references, or clone it if there are. At this
+        // point there should never be any other references; we only clone this
+        // Arc when serializing the ringbuf to a `serde_json::Value`.
+        let ringbuf = Arc::make_mut(&mut self.request_debug_ringbuf);
+        let to_delete = (ringbuf.len() + requests.len())
+            .saturating_sub(MAX_REQUEST_DEBUG_BUFFER_LEN);
+        ringbuf.drain(0..to_delete);
+
+        requests.sort();
+        ringbuf.extend(requests);
+    }
+
+    async fn list_sleds(&self, opctx: &OpContext) -> Result<Vec<Sled>> {
+        Ok(self
+            .datastore
+            .sled_list_all_batched(&opctx, SledFilter::TufArtifactReplication)
+            .await?
+            .into_iter()
+            .map(|sled| Sled {
+                id: SledUuid::from_untyped_uuid(sled.identity.id),
+                client: sled_client_from_address(
+                    sled.identity.id,
+                    sled.address(),
+                    &opctx.log,
+                ),
+                depot_base_url: format!(
+                    "http://{}",
+                    sled.address_with_port(sled.repo_depot_port.into())
+                ),
+            })
+            .collect())
+    }
+
+    async fn list_artifacts_from_database(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<Inventory> {
+        let mut inventory = Inventory::default();
+        let mut paginator = Paginator::new(SQL_BATCH_SIZE);
+        while let Some(p) = paginator.next() {
+            let batch = self
+                .datastore
+                .update_tuf_artifact_list(opctx, &p.current_pagparams())
+                .await?;
+            paginator = p.found_batch(&batch, &|a| a.id.into_untyped_uuid());
+            for artifact in batch {
+                // Two artifacts can have the same sha256 but all the values in
+                // this insertion step are the default value.
```

Comment:
```suggestion
                // Two artifacts can have the same sha256.  In that case, this will clobber what's there, but that's okay because all the values inserted up to this point are the default value.
```

(or just replace the code with `inventory.0.entry(artifact.sha256.0).or_insert_with(...)`, which I think communicates that the author knows the thing might already be there)